src/internal/storage/chunk/client.go

```diff
@@ -69,44 +96,51 @@ func (c *trackedClient) Create(ctx context.Context, md Metadata, chunkData []byt
 			return err
 		}
 		if len(ents) > 0 {
-			needUpload = false
+			needUpload = boolPtr(false)
 			return nil
 		}
-		if err := tx.Get(&gen, `
+		var gen2 uint64
+		if err := tx.Get(&gen2, `
 		INSERT INTO storage.chunk_objects (chunk_id, size)
 		VALUES ($1, $2)
 		RETURNING gen
 		`, chunkID, md.Size); err != nil {
 			return err
 		}
-		needUpload = true
+		needUpload = boolPtr(true)
+		gen = &gen2
 		return nil
 	}); err != nil {
-		return nil, err
-	}
-	if err := c.renewer.Add(ctx, chunkID); err != nil {
-		return nil, err
-	}
-	if !needUpload {
-		return chunkID, nil
-	}
-	key := chunkKey(chunkID, gen)
-	if err := c.store.Put(ctx, key, chunkData); err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-	_, err := c.db.ExecContext(ctx, `
+	return needUpload, gen, nil
+}
+
+// afterUpload marks the (chunkID, gen) pair as a successfully uploaded object in postgres.
+func (c *trackedClient) afterUpload(ctx context.Context, chunkID ID, gen uint64) error {
+	res, err := c.db.ExecContext(ctx, `
 	UPDATE storage.chunk_objects
 	SET uploaded = TRUE
 	WHERE chunk_id = $1 AND gen = $2
 	`, chunkID, gen)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	return chunkID, nil
+	affected, err := res.RowsAffected()
+	if err != nil {
+		return err
+	}
+	if affected < 1 {
+		return errors.Errorf("no chunk entry for object post upload: chunk=%v gen=%v", chunkID, gen)
```

Comment:
It seems like this error and the one below should have the same outcome. Either both return an error or panic. These would be pretty serious errors so maybe a panic makes sense, but taking down the whole process may be extreme.

Reply:
This first error (line 134) could happen if there was a long pause when uploading a chunk.  It shouldn't take the whole process down.  Affecting more than one, would be a logic error.  It means postgres has an incorrect primary key, or something serious enough to want to prevent any database activity.